src/meta/duplication/meta_duplication_service.cpp

```diff
@@ -184,50 +209,97 @@ void meta_duplication_service::add_duplication(duplication_add_rpc rpc)
+                break;
+            }
+        }
+
+        if (dup) {
+            // The duplication for the same app to the same remote cluster has existed.
+            remote_app_name = dup->remote_app_name;
+            LOG_INFO("no need to add duplication, since it has existed: app_name={}, "
+                     "remote_cluster_name={}, remote_app_name={}",
+                     request.app_name,
+                     request.remote_cluster_name,
+                     remote_app_name);
+        } else {
+            // Check if other apps of this cluster are duplicated to the same remote app.
+            for (const auto &kv : _state->_exist_apps) {
+                if (kv.first == request.app_name) {
+                    // Skip this app since we want ot check other apps.
+                    continue;
+                }
+
+                for (const auto &ent : kv.second->duplications) {
+                    if (ent.second->remote_cluster_name != request.remote_cluster_name) {
+                        continue;
+                    }
+
+                    if (ent.second->remote_app_name != remote_app_name) {
+                        continue;
+                    }
+
+                    response.err = ERR_INVALID_PARAMETERS;
+                    LOG_WARNING_DUP_HINT_AND_RETURN(response,
+                                                    "illegal operation: another app({}) is also "
+                                                    "duplicated to the same remote app("
+                                                    "cluster={}, app={})",
+                                                    kv.first,
+                                                    request.remote_cluster_name,
+                                                    remote_app_name);
+                }
+            }
         }
     }
+
     if (!dup) {
-        dup = new_dup_from_init(request.remote_cluster_name, std::move(meta_list), app);
+        dup = new_dup_from_init(
+            request.remote_cluster_name, remote_app_name, std::move(meta_list), app);
     }
-    do_add_duplication(app, dup, rpc);
+
+    do_add_duplication(app, dup, rpc, remote_app_name);
 }
 
 // ThreadPool(WRITE): THREAD_POOL_META_STATE
 void meta_duplication_service::do_add_duplication(std::shared_ptr<app_state> &app,
                                                   duplication_info_s_ptr &dup,
-                                                  duplication_add_rpc &rpc)
+                                                  duplication_add_rpc &rpc,
+                                                  const std::string &remote_app_name)
 {
-    const auto err = dup->start(rpc.request().is_duplicating_checkpoint);
+    const auto &err = dup->start(rpc.request().is_duplicating_checkpoint);
     if (dsn_unlikely(err != ERR_OK)) {
-        LOG_ERROR("start dup[{}({})] failed: err = {}", app->app_name, dup->id, err);
-        return;
+        LOG_ERROR_DUP_HINT_AND_RETURN(
```

Comment:
Ditto.

The `return` word is missing seems a little confused although it's in the macro name.